File: src/main/java/io/nayasis/sample/registrylock/redis/core/redis/Locker.java
package io.nayasis.sample.registrylock.redis.core.redis;

import org.springframework.data.redis.connection.RedisConnectionFactory;
import org.springframework.integration.redis.util.RedisLockRegistry;
import org.springframework.integration.support.locks.LockRegistry;

import java.util.concurrent.TimeUnit;
import java.util.concurrent.locks.Lock;

public class Locker {

    private LockRegistry lockRegistry;
    private String       registryKey;

    public Locker( RedisConnectionFactory connectionFactory, String registryKey ) {
        this.registryKey  = registryKey;
        this.lockRegistry = new RedisLockRegistry( connectionFactory, registryKey, 10_000 );
    }

    public boolean lock( String lockKey, Runnable runnable ) {

        Lock lock = lockRegistry.obtain( lockKey );

        try {

            boolean acquired = lock.tryLock( 10, TimeUnit.SECONDS );

            if( acquired ) {
                runnable.run();
                return true;
            }
            return false;

        } catch ( InterruptedException e ) {
            Thread.currentThread().interrupt();
            throw new RuntimeException( String.format( "Interrupted in locking [%s]", registryKey) );
        } finally {
            lock.unlock();
        }

    }

}

File: src/main/kotlin/com/github/nayasis/sample/distributedlock/lock/Locker.kt
@file:Suppress("SpringJavaInjectionPointsAutowiringInspection")

package com.github.nayasis.sample.distributedlock.lock

import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Configuration
import org.springframework.integration.support.locks.LockRegistry
import org.springframework.stereotype.Component
import java.lang.Thread.sleep
import java.util.concurrent.TimeUnit
import java.util.concurrent.locks.Lock

@Component
class Locker(
    val lockRegistry: LockRegistry,
    @Value("\${distributed-lock.timeout.connection:1000}")
    val connectionTimeout: Long,
    @Value("\${distributed-lock.retry.count:10}")
    val retryCount: Int,
    @Value("\${distributed-lock.retry.sleep:200}")
    val retrySleep: Long,
) {

    fun <T> lock(lockKey:String, fn:() -> T): T {
        val lock = tryLock(lockKey)
        try {
            return fn()
        } finally {
            try { lock.unlock() } catch (e: Exception) {}
        }
    }

    private fun tryLock(lockKey: String): Lock {
        val lock = lockRegistry.obtain(lockKey)
        for( i in 0..retryCount ) {
            if( i != 0 )
                sleep( retrySleep )
            if( lock.tryLock(connectionTimeout, TimeUnit.MILLISECONDS) )
                return lock
        }
        throw TimeoutException("Fail to acquire lock.")
    }

}